src/EditorFeatures/Core/Workspaces/AbstractTextBufferVisibilityTracker.cs

```diff
@@ -0,0 +1,227 @@
+            {
+                data = new VisibleTrackerData(this, subjectBuffer);
+                _subjectBufferToCallbacks.Add(subjectBuffer, data);
+            }
+
+            data.AddCallback(callback);
+        }
+
+        public void UnregisterForVisibilityChanges(ITextBuffer subjectBuffer, Action callback)
+        {
+            _threadingContext.ThrowIfNotOnUIThread();
+
+            // Both of these methods must succeed.  Otherwise we're somehow unregistering something we don't know about.
+            Contract.ThrowIfFalse(_subjectBufferToCallbacks.TryGetValue(subjectBuffer, out var data));
+            Contract.ThrowIfFalse(data.Callbacks.Contains(callback));
+            data.RemoveCallback(callback);
+
+            // If we have nothing that wants to listen to information about this buffer anymore, then disconnect it
+            // from all events and remove our map.
+            if (data.Callbacks.Count == 0)
+            {
+                data.Dispose();
+                _subjectBufferToCallbacks.Remove(subjectBuffer);
+            }
+        }
+
+        public TestAccessor GetTestAccessor()
+            => new(this);
+
+        public readonly struct TestAccessor
+        {
+            private readonly AbstractTextBufferVisibilityTracker<TTextView, TVisibilityChangedCallback> _visibilityTracker;
+
+            public TestAccessor(AbstractTextBufferVisibilityTracker<TTextView, TVisibilityChangedCallback> visibilityTracker)
+            {
+                _visibilityTracker = visibilityTracker;
+            }
+
+            public void TriggerCallbacks(ITextBuffer subjectBuffer)
+            {
+                var data = _visibilityTracker._subjectBufferToCallbacks[subjectBuffer];
+                data.TriggerCallbacks();
+            }
+        }
+
+        protected sealed class VisibleTrackerData : IDisposable
+        {
+            public readonly HashSet<ITextView> TextViews = new();
+
+            private readonly AbstractTextBufferVisibilityTracker<TTextView, TVisibilityChangedCallback> _tracker;
+            private readonly ITextBuffer _subjectBuffer;
+            private readonly TVisibilityChangedCallback _visibilityChangedCallback;
+
+            /// <summary>
+            /// The callbacks that want to be notified when our <see cref="TextViews"/> change visibility.  Stored as an
+            /// <see cref="ImmutableHashSet{T}"/> so we can enumerate it safely without it changing underneath us.
+            /// </summary>
+            public ImmutableHashSet<Action> Callbacks { get; private set; } = ImmutableHashSet<Action>.Empty;
+
+            public VisibleTrackerData(
+                AbstractTextBufferVisibilityTracker<TTextView, TVisibilityChangedCallback> tracker,
+                ITextBuffer subjectBuffer)
+            {
+                _tracker = tracker;
+                _subjectBuffer = subjectBuffer;
+                _visibilityChangedCallback = tracker.GetVisiblityChangeCallback(this);
```

Comment:
defer to platform specific code to get the callback with which to register to visibility changed event.